[WPE][WebDriver] Implement automation keyboard and mouse input handlers for new WPE API

lauromoura opened the pull request:
#### 4a97b2ece9b549de56c9e960f11009ecc00967da
<pre>
[WPE][WebDriver] Implement automation keyboard and mouse input handlers for new WPE API
<a href="https://bugs.webkit.org/show_bug.cgi?id=273750">https://bugs.webkit.org/show_bug.cgi?id=273750</a>

Reviewed by Carlos Garcia Campos.

Using new WPEEvent class when we&apos;re under the new WPE API.

This commit also moves the WPE-related WebAutomationSession platformSimulate* methods
implementation to the new API file, calling the older libwpe-based API in the `LibWPE.cpp`
file if required.

* Source/WebKit/SourcesWPE.txt: Add new file
* Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
* Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.h: Added
* Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
* Source/WebKit/WPEPlatform/wpe/WPEKeymapXKB.cpp: Fix typo

Canonical link: <a href="https://commits.webkit.org/279376@main">https://commits.webkit.org/279376@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/398d7ce4aee56f7be26452534dae3f66f4ff3ea1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53243 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32590 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5734 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56522 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3969 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39954 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3709 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43160 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2583 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55341 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30736 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45970 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24291 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27666 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2125 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49432 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3454 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58118 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28388 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3429 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50561 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29605 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46190 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49879 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11622 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30526 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29366 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
```diff
@@ -30,6 +30,10 @@
+#if ENABLE(WPE_PLATFORM)
+#include <wpe/wpe-platform.h>
+#endif
+
```
Even if we have to duplicate some code, like virtual key conversion, I think it's clearer to move the new api implementation to its own file as Source/WebKit/UIProcess/Automation/wpe/WebAutomationSessionWPE.cpp

lauromoura replied:
Moved the new API implementation to the suggested file, wrapped in `platformSimulate*WPEPlatform` functions called from the method overrides already existing in the `LibWPE.cpp` file.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
```diff
@@ -81,25 +85,113 @@ static unsigned stateModifierForWPEButton(unsigned button)
+    if (modifiers & wpe_input_keyboard_modifier_alt)
+        newModifiers |= WPE_MODIFIER_KEYBOARD_ALT;
+    if (modifiers & wpe_input_keyboard_modifier_meta)
+        newModifiers |= WPE_MODIFIER_KEYBOARD_META;
```
We should not use the old api values here, what we need is a new modifiersToEventState to use new api modifier values.

lauromoura replied:
Reverted this change, replaced by implementing the new api version as static methods on its own file.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
```diff
@@ -81,25 +85,113 @@ static unsigned stateModifierForWPEButton(unsigned button)
+    if (modifiers & wpe_input_pointer_modifier_button4)
+        newModifiers |= WPE_MODIFIER_POINTER_BUTTON4;
+    if (modifiers & wpe_input_pointer_modifier_button5)
+        newModifiers |= WPE_MODIFIER_POINTER_BUTTON5;
```
Same here,  but button modifiers are already handled by the other conversion helper no?

lauromoura replied:
Ditto.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
```diff
@@ -81,25 +85,113 @@ static unsigned stateModifierForWPEButton(unsigned button)
+{
+#if ENABLE(WPE_PLATFORM)
+    if (auto* view = page.wpeView()) {
+        unsigned wpeModifiers = newAPIModifiersFromRaw(modifiers);
```
What you call raw here is actually the old api modifiers.

lauromoura replied:
Ditto.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
```diff
@@ -81,25 +85,113 @@ static unsigned stateModifierForWPEButton(unsigned button)
+        auto* event = wpe_event_pointer_button_new(state ? WPE_EVENT_POINTER_DOWN : WPE_EVENT_POINTER_UP, view, WPE_INPUT_SOURCE_MOUSE,
+            0, static_cast<WPEModifiers>(wpeModifiers), button, location.x(), location.y(), pressCount);
+        wpe_view_event(view, event);
+        wpe_event_unref(event);
```
If you include GRefPtrWPE.h you can use GRefPtr here for the event.

lauromoura replied:
Good catch. Thanks!

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
```diff
@@ -176,6 +269,50 @@ static void doKeyStrokeEvent(struct wpe_view_backend* viewBackend, bool pressed,
+#if ENABLE(WPE_PLATFORM)
+    if (auto* view = page.wpeView()) {
+
+        auto* display = wpe_display_get_default();
```
WE should not use the default display here, use wpe_view_get_display instead

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
```diff
@@ -176,6 +269,50 @@ static void doKeyStrokeEvent(struct wpe_view_backend* viewBackend, bool pressed,
+            LOG(Automation, "WebAutomationSession::doKeyStrokeEvent: Failed to find WPE display. Ignoring event.");
+            return;
+        }
+        GError* error = nullptr;
```
Use GUniqueOutPtr

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
```diff
@@ -176,6 +269,50 @@ static void doKeyStrokeEvent(struct wpe_view_backend* viewBackend, bool pressed,
+        auto* xkbState = wpe_keymap_xkb_get_xkb_state(keymap);
+        unsigned keyval = xkb_state_key_get_one_sym(xkbState, keyCode);
+        if (keyval == XKB_KEY_NoSymbol)
+            return;
```
We should add wpe_keymap_get_entries_for_keycode() to the new API, so that you don't need to deal with xkb

carlosgcampos replied:
Sorry, I meant wpe_keymap_translate_keyboard_state or similar.

lauromoura replied:
I removed the xkb specific code by using `wpe_keymap_get_entries_for_keyval`, the new API equivalent of `wpe_input_xkb_context_get_entries_for_key_code`.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
```diff
@@ -134,10 +141,27 @@ void WebAutomationSession::platformSimulateMouseInteraction(WebPageProxy& page,
 
+// Called by platform-indendent code to convert the current platform-dependent raw modifiers into generic WebEventModifiers,
+// which will be passed to the platformSimulateFooBarInteraction methods.
 OptionSet<WebEventModifier> WebAutomationSession::platformWebModifiersFromRaw(unsigned modifiers)
```
We need to change this to receive a page too, even if it's unused in other platforms, because the given modifiers contains values from the old api when using the old api and from the new one when using the new api. So we need to decide like in the other cases.

lauromoura replied:
Changed in the new iteration, editing the GTK and Mac calls accordingly.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
```diff
@@ -134,10 +141,27 @@ void WebAutomationSession::platformSimulateMouseInteraction(WebPageProxy& page,
+    if (modifiers & WPE_MODIFIER_KEYBOARD_SHIFT)
+        webModifiers.add(WebEventModifier::ShiftKey);
+    // WPE has no Caps Lock modifier.
+#else
```
This is not a build time decision.

lauromoura replied:
Updated, leaving only the `WPE_PLATFORM` guards around the new API code and replacing the `#else` with a normal runtime `if` check whether we're running under the new API or not.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
```diff
@@ -104,10 +105,16 @@ void WebAutomationSession::platformSimulateMouseInteraction(WebPageProxy& page,
+        platformSimulateMouseInteractionWPEPlatform(page, interaction, button, locationInView, keyModifiers, pointerType, m_currentModifiers);
+        return;
+    }
+#endif
```
It seems you are doing the device scale conversion again in `platformSimulateMouseInteractionWPEPlatform`. Since the future is the new api, I would do this in the opposite way. The entry points in the new WPE file, calling the old api impl when page doesn't have a wpe view.

lauromoura replied:
Moved the entry points to the `WPE.cpp` file, refactoring the existing code into `FooLibWPE(...)` functions to be called from there.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,415 @@
+    if (!display) {
+        LOG(Automation, "WebAutomationSession::doKeyStrokeEvent: Failed to find WPE display. Ignoring event.");
+        return;
+    }
```
`wpe_view_get_display` never returns nullptr

lauromoura replied:
Updated.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,415 @@
+    // Note: wpe_event_keyboard_new uses the name 'keycode' for the value with the hardware key code
+    // and 'keyval' for the value with the translated value of the key (e.g. ascii-like).
+    // In comparison to the usage of this function in Wayland callbacks, here we have the variable names swapped,
+    // with WebAutomationSession using `keyCode` for the translated value and `keyval` for the XKB key code.
```
This is not what we want. First let's rename variables to avoid confusion, using always keyvale for the key symbol and keycode for the hardware code. What this function receives is a keyval, that we need to translate, so the first thing we need is to get the keycode associated the keyval using wpe_keymap_get_entries_for_keyval() as you are doing. Then we need to translate the keycode + modifiers into the actual keyval (so for example a + Shift ends up being A). For that we need new API in WPE platform, see https://github.com/WebKit/WebKit/pull/28524

lauromoura replied:
Done. Indeed, now if we send `key_down(SHIFT).key_down('e').perform()` we get an `E` and no modifiers instead of `e` + modifier.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,415 @@
+    }
+}
+
+// The same keysyms are shared between libwpe and the new API, so we don't to duplicate these
```
```suggestion
// The same keysyms are shared between libwpe and the new API, so we don't need to duplicate these
```

carlosgcampos replied:
But we are duplicating it no?

lauromoura replied:
Removed this comment.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,415 @@
+
+void platformSimulateKeyboardInteractionWPEPlatform(WebPageProxy& page, KeyboardInteraction interaction, std::variant<VirtualKey, CharKey>&& key, unsigned& currentModifiers)
+{
+    uint32_t keyCode;
```
Let's call this keyval

lauromoura replied:
Done.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,415 @@
+    return 0;
+}
+
+static uint32_t modifiersForKeyCode(unsigned keyCode)
```
This should be `modifiersForKeyval`

lauromoura replied:
Done.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,484 @@
+            webModifiers.add(WebEventModifier::ShiftKey);
+        if (modifiers & WPE_MODIFIER_KEYBOARD_CAPS_LOCK)
+            webModifiers.add(WebEventModifier::CapsLockKey);
+    } else {
```
We can just return here, and we don't need the else, nor the `#if` just for the `}`

lauromoura replied:
Done

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,484 @@
+static uint32_t modifiersToEventState(OptionSet<WebEventModifier> modifiers)
+{
+    uint32_t state = 0;
+    if (modifiers.contains(WebEventModifier::ControlKey))
+        state |= WPE_MODIFIER_KEYBOARD_CONTROL;
+    if (modifiers.contains(WebEventModifier::ShiftKey))
+        state |= WPE_MODIFIER_KEYBOARD_SHIFT;
+    if (modifiers.contains(WebEventModifier::AltKey))
+        state |= WPE_MODIFIER_KEYBOARD_ALT;
```
New api supports caps lock.

lauromoura replied:
Added capslock and meta modifiers for new api

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,484 @@
+#if ENABLE(WEBDRIVER_MOUSE_INTERACTIONS)
+
+#if ENABLE(WPE_PLATFORM)
+static void doMouseEvent(WebPageProxy& page, const WebCore::IntPoint& location, unsigned button, unsigned state, uint32_t modifiers)
```
This is new api specific, instead of unsigned state we could receive an event type, or if we want to be more specific to mouse events, it could be bool isPressed or something like that

lauromoura replied:
Went with `bool isPressed`.

carlosgcampos replied:
I still see unsigned state, here

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,484 @@
+        modifiers &= ~buttonModifiers;
+    unsigned pressCount = state ? wpe_view_compute_press_count(view, location.x(), location.y(), button, 0) : 0;
+    GRefPtr<WPEEvent> event = wpe_event_pointer_button_new(state ? WPE_EVENT_POINTER_DOWN : WPE_EVENT_POINTER_UP, view, WPE_INPUT_SOURCE_MOUSE,
+        0, static_cast<WPEModifiers>(modifiers), button, location.x(), location.y(), pressCount);
```
```suggestion
    GRefPtr<WPEEvent> event = adoptGRef(wpe_event_pointer_button_new(state ? WPE_EVENT_POINTER_DOWN : WPE_EVENT_POINTER_UP, view, WPE_INPUT_SOURCE_MOUSE,
        0, static_cast<WPEModifiers>(modifiers), button, location.x(), location.y(), pressCount));
```

lauromoura replied:
Nice catch. Thanks!

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,484 @@
+static void doMotionEvent(WebPageProxy& page, const WebCore::IntPoint& location, uint32_t modifiers)
+{
+    auto* view = page.wpeView();
+    GRefPtr<WPEEvent> event = wpe_event_pointer_move_new(WPE_EVENT_POINTER_MOVE, view, WPE_INPUT_SOURCE_MOUSE, 0, static_cast<WPEModifiers>(modifiers), location.x(), location.y(), 0, 0);
```
```suggestion
    GRefPtr<WPEEvent> event = adoptGRef(wpe_event_pointer_move_new(WPE_EVENT_POINTER_MOVE, view, WPE_INPUT_SOURCE_MOUSE, 0, static_cast<WPEModifiers>(modifiers), location.x(), location.y(), 0, 0));
```

lauromoura replied:
Ditto

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,484 @@
+    auto remainingModifiers = static_cast<WPEModifiers>(modifiers & ~consumedModifiers);
+
+    GRefPtr<WPEEvent> event = wpe_event_keyboard_new(pressed ? WPE_EVENT_KEYBOARD_KEY_DOWN : WPE_EVENT_KEYBOARD_KEY_UP, view, WPE_INPUT_SOURCE_KEYBOARD, 0,
+    remainingModifiers, keyCode, keyVal);
```
```suggestion
    GRefPtr<WPEEvent> event = adoptGRef(wpe_event_keyboard_new(pressed ? WPE_EVENT_KEYBOARD_KEY_DOWN : WPE_EVENT_KEYBOARD_KEY_UP, view, WPE_INPUT_SOURCE_KEYBOARD, 0,
        remainingModifiers, keyCode, keyVal));
```

lauromoura replied:
Ditto

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,484 @@
+    wpe_view_event(view, event.get());
+
+    if (doReleaseAfterPress) {
+        event = wpe_event_keyboard_new(WPE_EVENT_KEYBOARD_KEY_UP, view, WPE_INPUT_SOURCE_KEYBOARD, 0, static_cast<WPEModifiers>(modifiers), keyCode, keyVal);
```
```suggestion
        event = adoptGRef(wpe_event_keyboard_new(WPE_EVENT_KEYBOARD_KEY_UP, view, WPE_INPUT_SOURCE_KEYBOARD, 0, static_cast<WPEModifiers>(modifiers), keyCode, keyVal));
```

lauromoura replied:
Ditto

carlosgcampos commented on Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
```diff
@@ -0,0 +1,484 @@
+
+#if ENABLE(WPE_PLATFORM)
+    auto* view = page.wpeView();
+    GRefPtr<WPEEvent> event = wpe_event_scroll_new(view, WPE_INPUT_SOURCE_MOUSE, 0, static_cast<WPEModifiers>(0), delta.width(), delta.height(), false, false, location.x(), location.y());
```
```suggestion
    GRefPtr<WPEEvent> event = adoptGRef(wpe_event_scroll_new(view, WPE_INPUT_SOURCE_MOUSE, 0, static_cast<WPEModifiers>(0), delta.width(), delta.height(), false, false, location.x(), location.y()));
```

lauromoura replied:
Ditto

carlosgcampos reviewed (approved):
This looks good to me now, just fix the memory leaks before landing, please.

lauromoura committed:
[WPE][WebDriver] Implement automation keyboard and mouse input handlers for new WPE API
https://bugs.webkit.org/show_bug.cgi?id=273750

Reviewed by Carlos Garcia Campos.

Using new WPEEvent class when we're under the new WPE API.

This commit also moves the WPE-related WebAutomationSession platformSimulate* methods
implementation to the new API file, calling the older libwpe-based API in the `LibWPE.cpp`
file if required.

* Source/WebKit/SourcesWPE.txt: Add new file
* Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
* Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.h: Added
* Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
* Source/WebKit/WPEPlatform/wpe/WPEKeymapXKB.cpp: Fix typo

Canonical link: https://commits.webkit.org/279376@main
